rsocket-transport-netty/src/main/java/io/rsocket/transport/netty/server/WebsocketServerTransport.java

```diff
@@ -147,27 +161,92 @@ public void setTransportHeaders(Supplier<Map<String, String>> transportHeaders)
-                      FRAME_LENGTH_MASK,
-                      (in, out) -> {
-                        DuplexConnection connection =
-                            new WebsocketDuplexConnection((Connection) in);
-                        if (mtu > 0) {
-                          connection =
-                              new FragmentationDuplexConnection(
-                                  connection, ByteBufAllocator.DEFAULT, mtu, false, "server");
-                        }
-                        return acceptor.apply(connection).then(out.neverComplete());
-                      });
-                })
-            .bind()
-            .map(CloseableChannel::new);
+
+    if (isError != null) {
+      return isError;
+    }
+
+    if (headersPredicate != null) {
+      return server
+          .handle(
+              (request, response) -> {
+                transportHeaders.get().forEach(response::addHeader);
+                return response.sendWebsocket(
+                    null,
+                    FRAME_LENGTH_MASK,
+                    (in, out) -> {
+                      HttpHeaders headers = in.headers();
+                      if (!headersPredicate.test(headers)) {
+                        final WebSocketCloseStatus status =
+                            webSocketCloseStatusSupplier.apply(headers);
+                        return out.sendClose(status.code(), status.reasonText());
+                      }
+
+                      DuplexConnection connection = new WebsocketDuplexConnection((Connection) in);
+                      if (mtu > 0) {
+                        connection =
+                            new FragmentationDuplexConnection(
+                                connection, ByteBufAllocator.DEFAULT, mtu, false, "server");
+                      }
+                      return acceptor.apply(connection).then(out.neverComplete());
+                    });
+              })
+          .bind()
+          .map(CloseableChannel::new);
+    }
+
+    return server
+        .handle(
+            (request, response) -> {
+              transportHeaders.get().forEach(response::addHeader);
+              return response.sendWebsocket(
+                  null,
+                  FRAME_LENGTH_MASK,
+                  (in, out) -> {
+                    DuplexConnection connection = new WebsocketDuplexConnection((Connection) in);
+                    if (mtu > 0) {
+                      connection =
+                          new FragmentationDuplexConnection(
+                              connection, ByteBufAllocator.DEFAULT, mtu, false, "server");
+                    }
+                    return acceptor.apply(connection).then(out.neverComplete());
+                  });
+            })
+        .bind()
+        .map(CloseableChannel::new);
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+
+    private Predicate<HttpHeaders> headersPredicate;
```

Comment:
Why split headersPredicate and webSocketCloseStatusSupplier?  Seems like this could be combined into a single function call instead?

Reply:
Do you think that returning `null` in case everything is fine is a better approach? I was thinking about API which allows a user to filter incoming connection out and in case it is `rejected` to provide a specific close status. One of the options is to use SynchronousSink (as it is in Reactor) but it seems way complex as of normal developer standpoint (e.g. calling error or success)

Reply:
In this case splitting it means possibly evaluating the rules twice, so probably.  Or use optional?